rate-limit/src/lib.rs

```diff
@@ -0,0 +1,436 @@
+//! # Rate Limit
+//! A module to provide rate limit for arbitrary type Key and integer type Value
+//!
+//! - [`Config`](./trait.Config.html)
+//! - [`Call`](./enum.Call.html)
+//! - [`Module`](./struct.Module.html)
+//!
+//! ## Overview
+//!
+//! This module is a utility to provide rate limiter for arbitrary type Key and
+//! integer type Value, which can config limit rule to produce quota and consume
+//! quota, and expose quota consuming checking and whitelist that can bypass
+//! checks.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+#![allow(clippy::unused_unit)]
+
+use codec::MaxEncodedLen;
+use frame_support::{pallet_prelude::*, traits::UnixTime, transactional, BoundedVec};
+use frame_system::pallet_prelude::*;
+use orml_traits::{RateLimiter, RateLimiterError};
+use orml_utilities::OrderedSet;
+use scale_info::TypeInfo;
+use sp_runtime::traits::{SaturatedConversion, Zero};
+use sp_std::{prelude::*, vec::Vec};
+
+pub use module::*;
+pub use weights::WeightInfo;
+
+mod mock;
+mod tests;
+pub mod weights;
+
+#[frame_support::pallet]
+pub mod module {
+	use super::*;
+
+	/// Limit rules type.
+	#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+	pub enum RateLimitRule {
+		/// Each `blocks_count` blocks to reset remainer quota to `quota`
+		/// amount. is_allowed check return true when the remainer quota gte the
+		/// consume amount.
+		PerBlocks { blocks_count: u64, quota: u128 },
+		/// Each `secs_count` seconds to reset remainer quota to `quota` amount.
+		/// is_allowed check return true when the remainer quota gte the consume
+		/// amount.
+		PerSeconds { secs_count: u64, quota: u128 },
+		/// Each `blocks_count` blocks to increase `quota_increment` amount to
+		/// remainer quota and keep remainer quota lte `max_quota`. is_allowed
+		/// check return true when the remainer quota gte the consume amount.
+		TokenBucket {
+			blocks_count: u64,
+			quota_increment: u128,
+			max_quota: u128,
+		},
+		/// is_allowed check return true always.
+		Unlimited,
+		/// is_allowed check return false always.
+		NotAllowed,
+	}
+
+	/// The maximum length of KeyFilter inner key.
+	pub const MAX_FILTER_KEY_LENGTH: u32 = 256;
+
+	/// Match rules to fitler key is in bypass whitelist.
+	#[derive(PartialOrd, Ord, PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, TypeInfo, MaxEncodedLen)]
+	pub enum KeyFilter {
+		/// If the encoded key is equal to the vec, the key is in whitelist.
+		Match(BoundedVec<u8, ConstU32<MAX_FILTER_KEY_LENGTH>>),
+		/// If the encoded key starts with the vec, the key is in whitelist.
+		StartsWith(BoundedVec<u8, ConstU32<MAX_FILTER_KEY_LENGTH>>),
+		/// If the encoded key ends with the vec, the key is in whitelist.
+		EndsWith(BoundedVec<u8, ConstU32<MAX_FILTER_KEY_LENGTH>>),
+	}
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
```

Comment:
should also have a BlockNumberProvider so we can for example use relaychain block number if we want to